### PR TITLE
v0.3.11: allows json data requests

### PR DESCRIPTION
**Pull request by andywerner**

### CODE REVIEW

**Comment by amrsoll** [comment]
~Looking at it from the phone, it would 1st seem that indentation (space/tabs) wasn't respected~

File: octoprint_findmymrbeam/__init__.py [hunk 1]
```diff
@@ -139,10 +163,29 @@ def is_online_gif(self, secret):
...
 		# send a transparent 1x1 px gif
 		import base64
-		response = flask.make_response(bytes(base64.b64decode("R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7")))
+		response = flask.make_response(
+			bytes(base64.b64decode("R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7")))
 		response.headers["Content-Type"] = "image/gif"
 		return response
 
+	@octoprint.plugin.BlueprintPlugin.route("/<secret>.json", methods=["GET", "OPTIONS"])
+	def get_device_info(self, secret):
```
Comment (amrsoll) [inline octoprint_findmymrbeam/__init__.py]: Consider a different name, this does not seem to describe the function, or please add a docstring.

`get_device_info` is also used in the mrb plugin, doesn't seem to be related.

File: octoprint_findmymrbeam/__init__.py [hunk 2]
```diff
@@ -61,12 +60,45 @@ def __init__(self):
...
+					uuid=self._uuid,
+					search_id=self._search_id,
+					name=self._find_name(),
+					hostname=socket.gethostname(),
+					local_ips=self._get_local_ips(),
+					netconnectd_state=self._get_netconnectd_state(),
+					modes=self._get_internal_modes(),
+					query="plugin/{}/{}".format(self._identifier, self._secret),
+					plugin_version=self._get_plugin_version(),
+					)
```
Comment (amrsoll) [inline octoprint_findmymrbeam/__init__.py]: No difference between the functions? If there is a difference, please make it more obvious, like making a separate function that contains all the keys except the few keys these functions modify.

File: octoprint_findmymrbeam/__init__.py [hunk 3]
```diff
@@ -61,12 +60,45 @@ def __init__(self):
 		self._all_secrets = self._not_so_secret + [self._secret]
 
 	def initialize(self):
+		self._uuid = self._get_setting([["plugins", "discovery", "upnpUuid"], ], ["public", "uuid"])
+		if self._uuid is None:
+			self._generate_uuid()
+		self._search_id = self._settings.get(["public", "search_id"])
+		if self._search_id is None:
+			self._generate_search_id()
```
Comment (amrsoll) [inline octoprint_findmymrbeam/__init__.py]:
```
self._uuid = ... or ...
self._search_id = ... or ...
```
You should make the "generate" functions return the value instead of assigning it to the class attribute

File: octoprint_findmymrbeam/__init__.py [hunk 4]
```diff
@@ -366,52 +381,39 @@ def _get_netconnectd_state(self):
...
+				# 	version = pluginInfo.version
+				if pluginInfo.implementation._plugin_version:
+					self._mrb_plugin_version = pluginInfo.implementation._plugin_version
+			except Exception as e:
+				self._logger.exception(
+					"Exception while reading version from mrbeam: {}".format(e)
+				)
+		return self._mrb_plugin_version
+
+	def _perform_update_request(self):
```
Comment (amrsoll) [inline octoprint_findmymrbeam/__init__.py]: Too many `if ... is None` and `try: ... except ...`

```py
from itertools import compress
modes_bool = (pluginInfo.implementation.support_mode, pluginInfo.implementation.calibration_tool_mode,)
# return a tuple of activated modes like support or calibration tool
return tuple(compress((MODE_SUPPORT, MODE_CALIBRATION_TOOL), modes_bool))
```
These functions are too simple to warrant the `self._internal_modes` caching.

File: octoprint_findmymrbeam/static/js/findmymrbeam_settings_viewmodel.js [hunk 5]
```diff
@@ -61,7 +73,20 @@ $(function () {
             self.verifyByFrontend();
         };
 
-        self.verifyByFrontend = function() {
+        self.sendToExtension = function () {
+            if (self.enabled()) {
+                // send to find.mr-beam extension's contentScript
+                // TODO: should be done whenever searchID changes...
```
Comment (amrsoll) [inline octoprint_findmymrbeam/static/js/findmymrbeam_settings_viewmodel.js:79]: You could have `searchID` [subscribe] to `sendToExtension`

[subscribe]: https://knockoutjs.com/documentation/observables.html#explicitly-subscribing-to-observables

File: octoprint_findmymrbeam/static/js/findmymrbeam_settings_viewmodel.js [hunk 6]
```diff
@@ -6,32 +6,44 @@ $(function () {
...
+                + "&search_id=" + encodeURIComponent(self.searchId())
+                + "&public_ip=" + encodeURIComponent(self.public_ip())
+                + "&public_ip6=" + encodeURIComponent(self.public_ip6())
                 ;
         });
         self.verified = ko.observable(false);
         self.verification_response = null;
 
         self.onAllBound = function () {
+            self.enabled = self.settings.settings.plugins.findmymrbeam.enabled;
```
Comment (amrsoll) [inline octoprint_findmymrbeam/static/js/findmymrbeam_settings_viewmodel.js:31]: Wouldn't you disable the whole extension instead of having an "enabled" button?

File: octoprint_findmymrbeam/__init__.py [hunk 7]
```diff
@@ -222,8 +272,14 @@ def _find_name(self):
...
-	def _find_color(self):
-		return self._settings.global_get(["appearance", "color"])
+	def _get_local_ips(self):
+		local_ips = []
+		for addr in octoprint.util.interface_addresses():
+			if netaddr.IPAddress(addr) in LOCALHOST:
+				continue
+			# if addr not in ("10.250.250.1",):
+			local_ips.append(addr)
+		return local_ips
```
Comment (amrsoll) [inline octoprint_findmymrbeam/__init__.py]: Consider shortening
```py
return [ addr for addr in octoprint.util.interface_addresses() if netaddr.IPAddress(addr) not in LOCALHOST ]
# or with a filter
return list(filter(lambda x: netaddr.IPAddress(x) not in LOCALHOST, octoprint.util.interface_addresses()))
```

**Review by amrsoll — CHANGES_REQUESTED** [review]
It seems a bit rushed, I think it could benefit from some more readability :)

Extra tip: To know which modes the device is in (support / calibration), you could also use [helper functions] to better expose a function from the mrb plugin instead of going through the plugin info. The `_get_internal_modes` could be rewritten in the mrb plugin instead (it's probably already there)

[helper functions]: https://docs.octoprint.org/en/1.4.2/plugins/helpers.html

**Comment by andywerner** [comment]
Agree on the helper thing. But it's scope is too big for now.

**Comment by amrsoll** [comment]
> Too many `if ... is None` and `try: ... except ...`
> 
> ```python
> from itertools import compress
> modes_bool = (pluginInfo.implementation.support_mode, pluginInfo.implementation.calibration_tool_mode,)
> # return a tuple of activated modes like support or calibration tool
> return tuple(compress((MODE_SUPPORT, MODE_CALIBRATION_TOOL), modes_bool))
> ```
> 
> These functions are too simple to warrant the `self._internal_modes` caching.

I still think [this comment](https://github.com/mrbeam/OctoPrint-FindMyMrBeam/pull/27#discussion_r607594680) is relevant, but this is already nice :+1:

**Review by amrsoll — APPROVED** [review]
